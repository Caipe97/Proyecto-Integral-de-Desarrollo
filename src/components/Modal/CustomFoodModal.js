import React, { useState } from 'react';
import Modal  from 'react-bootstrap/Modal';
import { Button } from 'react-bootstrap';
import TextField from '@material-ui/core/TextField';
import Combobox from "react-widgets/Combobox";
import "react-widgets/styles.css";
import Edit from '../../images/edit.png';

const CustomFoodModal = (props) =>{
  const [modalShow, setModalShow] = useState(false);
  const [state, setState] = useState({
    name: '',
    recommendedServing: '',
    caloriesPerServing: '',
    foodCategoryId: '',
    message: ''
  })

  const handleChange = event => {
    setState({
      ...state,
      [event.target.name]: event.target.value
    })
  };

  const handleChangeComboBox = event => {
    setState({
      ...state,
      foodCategoryId: event.foodCategoryId
    })
  };

  if(props.edit){
    
    const handleSubmitEdit = async (event) => {
      event.preventDefault();
      const data = await props.onEditCustomFood(props.foodId, state.name, state.recommendedServing, state.caloriesPerServing, state.foodCategoryId);
      if(data){
        setState({
          ...state,
          message: 'Alimento editado exitosamente'
        });
      }
    };

    return (
      <>
      <img src={Edit} alt='editar' style={{width: '20px', height: '20px', cursor: 'pointer'}} onClick={() => {
        setState({
          name: props.row.name,
          recommendedServing: props.row.recommendedServing,
          caloriesPerServing: props.row.caloriesPerServing,
          foodCategoryId: props.row.foodCategoryId
        });
        setModalShow(true)
        }}/>
        
        <Modal
          show={modalShow}
          size="lg"
          aria-labelledby="contained-modal-title-vcenter"
          centered
        >
          <Modal.Header>
            <Modal.Title id="contained-modal-title-vcenter">
            {'Editar Alimento Personalizado'}
            </Modal.Title>
            <Button type="button" onClick={() => {setModalShow(false); setState({...state, message: ''});}} style={{ backgroundColor: 'white', borderColor: 'white', color: "black"}}>X</Button>
          </Modal.Header>
          <Modal.Body>
            <form style={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: '100%', flexDirection: 'column' }}>
              <TextField label="Nombre" name='name' type='name' value={state.name} onChange={handleChange} required />
              <TextField label="Porción Recomendada" name='recommendedServing' type='recommendedServing' value={state.recommendedServing} onChange={handleChange} required />
              <TextField label="Calorias por Porción" name='caloriesPerServing' type='caloriesPerServing' value={state.caloriesPerServing} onChange={handleChange} required />
              <Combobox
                data={props.foodCategories}
                textField='name'
                onSelect={handleChangeComboBox}
                groupBy={category => category.userId}
                renderListGroup={ ({group}) => ( //group es el userId
                  <span>{group === null ? 'Default' : 'Custom'}</span>
                )}
              />
              <button onClick={handleSubmitEdit} className='button'>Finalizar edición</button>
              <p>{state.message}</p>
            </form>
          </Modal.Body>
        </Modal>
      </>
    );
  } else {
    const handleSubmitAdd = async (event) => {
      event.preventDefault();
      const data = await props.onAddCustomFood(state.name, state.recommendedServing, state.caloriesPerServing, state.foodCategoryId, props.userId);
      if(data){
        setState({
          name: '',
          recommendedServing: '',
          caloriesPerServing: '',
          foodCategoryId: '',
          message: 'Alimento agregado exitosamente'
        });
      }
    };

    return (
      <>
        <Button  type="button" variant="primary" onClick={() => setModalShow(true)} style={{backgroundColor: 'rgb(18, 207, 90)', borderColor: 'rgb(18, 207, 90)'}}>
          Agregar Alimento Personalizado
        </Button>
        
        <Modal
          show={modalShow}
          size="lg"
          aria-labelledby="contained-modal-title-vcenter"
          centered
        >
          <Modal.Header>
            <Modal.Title id="contained-modal-title-vcenter">
            {'Agregar Alimento Personalizado'}
            </Modal.Title>
            <Button type="button" onClick={() => setModalShow(false)} style={{ backgroundColor: 'white', borderColor: 'white', color: "black"}}>X</Button>
          </Modal.Header>
          <Modal.Body>
            <form style={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: '100%', flexDirection: 'column' }}>
              <TextField label="Nombre" name='name' type='name' value={state.name} onChange={handleChange} required />
              <TextField label="Porción Recomendada" name='recommendedServing' type='recommendedServing' value={state.recommendedServing} onChange={handleChange} required />
              <TextField label="Calorias por Porción" name='caloriesPerServing' type='caloriesPerServing' value={state.caloriesPerServing} onChange={handleChange} required />
              <Combobox
                data={props.foodCategories}
                textField='name'
                onSelect={handleChangeComboBox}
                groupBy={category => category.userId}
                renderListGroup={ ({group}) => ( //group es el userId
                  <span>{group === null ? 'Default' : 'Custom'}</span>
                )}
              />
              <button onClick={handleSubmitAdd} className='button'>Agregar</button>
              <p>{state.message}</p>
            </form>

          </Modal.Body>
        </Modal>
      </>
    );
  }

}
  
  export default CustomFoodModal;