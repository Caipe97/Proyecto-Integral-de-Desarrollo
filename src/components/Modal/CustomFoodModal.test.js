import React from 'react';
import { shallow } from 'enzyme';
import CustomFoodModal from './CustomFoodModal';

describe('CustomFoodModal tests', () => {
    const preventDefault = { preventDefault: jest.fn() }
    let wrapper;
    let wrapper2;
    beforeEach(() => {
        let mockProps = {
            
            foodCategories: [
                {
                    foodCategoryId: 1,
                    name: "Fruta",
                    createdAt: "2021-09-30T14:09:34.756Z",
                    updatedAt: "2021-09-30T14:09:34.756Z",
                    userId: null
                },
                {
                    foodCategoryId: 7,
                    name: "Fideos",
                    createdAt: "2021-09-30T18:58:58.542Z",
                    updatedAt: "2021-09-30T18:58:58.542Z",
                    userId: null
                }
            ],
            preventDefault: jest.fn(),
            onAddCustomFood: jest.fn(),
            onEditCustomFood:jest.fn(),
            edit: false,
            userId: 1,
        };
        let mockProps2 = {
            row:{
                name: 'Milanesa', recommendedServing: 85, caloriesPerServing: 198, foodCategoryId:1
            },
            foodCategories: [
                {
                    foodCategoryId: 1,
                    name: "Fruta",
                    createdAt: "2021-09-30T14:09:34.756Z",
                    updatedAt: "2021-09-30T14:09:34.756Z",
                    userId: null
                },
                {
                    foodCategoryId: 7,
                    name: "Fideos",
                    createdAt: "2021-09-30T18:58:58.542Z",
                    updatedAt: "2021-09-30T18:58:58.542Z",
                    userId: null
                }
            ],
            preventDefault: jest.fn(),
            onAddCustomFood: jest.fn(),
            onEditCustomFood:jest.fn(),
            edit: true,
            userId: 1,
        };
        wrapper = shallow(<CustomFoodModal {...mockProps} />

        );
        wrapper2 = shallow(<CustomFoodModal {...mockProps2} />

        );
    })
    it('expect to render AddCategoryModal component to render when edit is false', () => {
        expect(wrapper).toMatchSnapshot();
    })
    it('expect to render AddCategoryModal component to render when edit is true', () => {
        expect(wrapper2).toMatchSnapshot();
    })
    it('handleChange should update the state correctly when writing on the inputs', () => {
        wrapper.find('[name="name"]').at(0).simulate('change', { target: { name: 'name', value: 'Patata' } });
        wrapper.find('[name="name"]').at(0).simulate('change', { target: { name: 'recommendedServing', value: 456 } });
        wrapper.find('[name="name"]').at(0).simulate('change', { target: { name: 'caloriesPerServing', value: 123 } });
    })
    //else props.edit
    it('On click if edit false, setModal for agregaralimentopersonalizado', () => {
        wrapper.find('[type="button"]').at(1).simulate('click', preventDefault);
    })
    it('HandleSubmitAdd on click ', () => {
        wrapper.find('[className="button"]').at(0).simulate('click', preventDefault);
    })
    it('On click if edit false, setModal for agregaralimentopersonalizado', () => {
        wrapper.find('[type="button"]').at(0).simulate('click', preventDefault);
    })
    it('handleSubmit expect t11', () => {
        wrapper2.find('[type="button"]').at(0).simulate('click', preventDefault);
    })
    it('handleSubmitEdit when edition finish ', async () => {
        wrapper2.find('button').at(0).simulate('click', preventDefault);
      
    })
    //if props.edit
    it('expect on click when change name,recommendedServing, caloriesPerServing and foodCategoryId',() => {
        wrapper2.find('[alt="editar"]').simulate('click');
    })
    it('HandleSubmitEdit on click', () => {
        wrapper2.find('[type="button"]').at(0).simulate('click', preventDefault);
    })
})
