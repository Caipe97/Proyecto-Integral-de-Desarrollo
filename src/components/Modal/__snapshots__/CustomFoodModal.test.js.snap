// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomFoodModal tests expect to render AddCategoryModal component to render when edit is false 1`] = `
<Fragment>
  <Button
    active={false}
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "rgb(18, 207, 90)",
        "borderColor": "rgb(18, 207, 90)",
      }
    }
    type="button"
    variant="primary"
  >
    Agregar Alimento Personalizado
  </Button>
  <Modal
    animation={true}
    aria-labelledby="contained-modal-title-vcenter"
    autoFocus={true}
    backdrop={true}
    centered={true}
    dialogAs={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    enforceFocus={true}
    keyboard={true}
    restoreFocus={true}
    show={false}
    size="lg"
  >
    <ModalHeader
      closeButton={false}
      closeLabel="Close"
    >
      <ModalTitle
        id="contained-modal-title-vcenter"
      >
        Agregar Alimento Personalizado
      </ModalTitle>
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "white",
            "color": "black",
          }
        }
        type="button"
        variant="primary"
      >
        X
      </Button>
    </ModalHeader>
    <ModalBody>
      <form
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
            "height": "100%",
            "justifyContent": "center",
          }
        }
      >
        <WithStyles(ForwardRef(TextField))
          label="Nombre"
          name="name"
          onChange={[Function]}
          required={true}
          type="name"
          value=""
        />
        <WithStyles(ForwardRef(TextField))
          label="Porción Recomendada"
          name="recommendedServing"
          onChange={[Function]}
          required={true}
          type="recommendedServing"
          value=""
        />
        <WithStyles(ForwardRef(TextField))
          label="Calorias por Porción"
          name="caloriesPerServing"
          onChange={[Function]}
          required={true}
          type="caloriesPerServing"
          value=""
        />
        <Combobox
          data={
            Array [
              Object {
                "createdAt": "2021-09-30T14:09:34.756Z",
                "foodCategoryId": 1,
                "name": "Fruta",
                "updatedAt": "2021-09-30T14:09:34.756Z",
                "userId": null,
              },
              Object {
                "createdAt": "2021-09-30T18:58:58.542Z",
                "foodCategoryId": 7,
                "name": "Fideos",
                "updatedAt": "2021-09-30T18:58:58.542Z",
                "userId": null,
              },
            ]
          }
          groupBy={[Function]}
          onSelect={[Function]}
          renderListGroup={[Function]}
          textField="name"
        />
        <button
          className="button"
          onClick={[Function]}
        >
          Agregar
        </button>
        <p />
      </form>
    </ModalBody>
  </Modal>
</Fragment>
`;

exports[`CustomFoodModal tests expect to render AddCategoryModal component to render when edit is true 1`] = `
<Fragment>
  <img
    alt="editar"
    onClick={[Function]}
    src="edit.png"
    style={
      Object {
        "cursor": "pointer",
        "height": "20px",
        "width": "20px",
      }
    }
  />
  <Modal
    animation={true}
    aria-labelledby="contained-modal-title-vcenter"
    autoFocus={true}
    backdrop={true}
    centered={true}
    dialogAs={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    enforceFocus={true}
    keyboard={true}
    restoreFocus={true}
    show={false}
    size="lg"
  >
    <ModalHeader
      closeButton={false}
      closeLabel="Close"
    >
      <ModalTitle
        id="contained-modal-title-vcenter"
      >
        Editar Alimento Personalizado
      </ModalTitle>
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "white",
            "color": "black",
          }
        }
        type="button"
        variant="primary"
      >
        X
      </Button>
    </ModalHeader>
    <ModalBody>
      <form
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
            "height": "100%",
            "justifyContent": "center",
          }
        }
      >
        <WithStyles(ForwardRef(TextField))
          label="Nombre"
          name="name"
          onChange={[Function]}
          required={true}
          type="name"
          value=""
        />
        <WithStyles(ForwardRef(TextField))
          label="Porción Recomendada"
          name="recommendedServing"
          onChange={[Function]}
          required={true}
          type="recommendedServing"
          value=""
        />
        <WithStyles(ForwardRef(TextField))
          label="Calorias por Porción"
          name="caloriesPerServing"
          onChange={[Function]}
          required={true}
          type="caloriesPerServing"
          value=""
        />
        <Combobox
          data={
            Array [
              Object {
                "createdAt": "2021-09-30T14:09:34.756Z",
                "foodCategoryId": 1,
                "name": "Fruta",
                "updatedAt": "2021-09-30T14:09:34.756Z",
                "userId": null,
              },
              Object {
                "createdAt": "2021-09-30T18:58:58.542Z",
                "foodCategoryId": 7,
                "name": "Fideos",
                "updatedAt": "2021-09-30T18:58:58.542Z",
                "userId": null,
              },
            ]
          }
          groupBy={[Function]}
          onSelect={[Function]}
          renderListGroup={[Function]}
          textField="name"
        />
        <button
          className="button"
          onClick={[Function]}
        >
          Finalizar edición
        </button>
        <p />
      </form>
    </ModalBody>
  </Modal>
</Fragment>
`;
